scvi/data/_utils.py

```diff
@@ -120,49 +121,43 @@ def _verify_and_correct_data_format(
         _set_data_in_registry(adata, data, attr_name, attr_key)
 
 
-def _make_obs_column_categorical(
-    adata,
-    column_key,
-    alternate_column_key,
-    categorical_dtype=None,
+def _make_column_categorical(
+    df: pd.DataFrame,
+    column_key: str,
+    alternate_column_key: str,
+    categorical_dtype: Optional[Union[str, CategoricalDtype]] = None,
 ):
     """
-    Makes the data in column_key in obs all categorical.
+    Makes the data in column_key in DataFrame all categorical.
 
-    Categorizes adata.obs[column_key], then saves category codes to
-    .obs[alternate_column_key] and the category mappings
-    to .uns["scvi"]["categorical_mappings"].
+    Categorizes df[column_key], then saves category codes to
+    df[alternate_column_key] and returns the category mappings.
     """
     if categorical_dtype is None:
-        categorical_obs = adata.obs[column_key].astype("category")
+        categorical_obs = df[column_key].astype("category")
     else:
-        categorical_obs = adata.obs[column_key].astype(categorical_dtype)
+        categorical_obs = df[column_key].astype(categorical_dtype)
 
     # put codes in .obs[alternate_column_key]
     codes = categorical_obs.cat.codes
     mapping = categorical_obs.cat.categories.to_numpy(copy=True)
     if -1 in np.unique(codes):
-        received_categories = adata.obs[column_key].astype("category").cat.categories
+        received_categories = df[column_key].astype("category").cat.categories
         raise ValueError(
             'Making .obs["{}"] categorical failed. Expected categories: {}. '
             "Received categories: {}. ".format(column_key, mapping, received_categories)
         )
-    adata.obs[alternate_column_key] = codes
+    df[alternate_column_key] = codes
 
     # make sure each category contains enough cells
-    unique, counts = np.unique(adata.obs[alternate_column_key], return_counts=True)
+    unique, counts = np.unique(getattr(df, alternate_column_key), return_counts=True)
```

Comment:
why getattr and not `df[alternate_column_key]`

Reply:
also wait isn't this call to np unique inefficient? We already called np unique on the codes earlier on 144?